AdblockPlusSafari/FilterList+Processing.m

```diff
@@ -0,0 +1,251 @@
+
++ (BOOL)parseExpiresString:(NSString *)expires
+                        to:(NSTimeInterval *)output
+{
+  NSError *error = nil;
+  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^(\\d+) days"
+                                                                         options:NSRegularExpressionCaseInsensitive
+                                                                           error:&error];
+  if (error != nil) {
+    return NO;
+  }
+  NSArray *matches = [regex matchesInString:expires
+                                    options:0
+                                      range:NSMakeRange(0, [expires length])];
+  NSAssert([matches count] <= 1, @"There should be at most one match");
+  NSTextCheckingResult *firstMatch = matches.firstObject;
+  if (firstMatch == nil) {
+    return NO;
+  }
+
+  *output = [[expires substringWithRange:[firstMatch range]] doubleValue] * 3600 * 24;
+  return YES;
+}
+
+// Get parser error
++ (NSError *)createParserError:(yajl_handle) hand
+{
+  NSUInteger errorBufferLength = 512;
+  uint8_t errorBuffer[errorBufferLength];
+  unsigned char *error_string = yajl_get_error(hand, 1, errorBuffer, errorBufferLength);
+  NSString *errorString = [[NSString alloc] initWithCString:(const char *)error_string encoding:NSASCIIStringEncoding];
+  return [NSError errorWithDomain:AdblockPlusErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: errorString}];
+}
+
+- (BOOL)parseFilterListFromURL:(NSURL *__nonnull)input
+                         error:(NSError *__nullable *__nonnull)error
+{
+  NSInputStream *inputStream = [NSInputStream inputStreamWithURL:input];
+
+  yajl_handle hand = NULL;
+  AdblockPlusProcessingContext *context = [[AdblockPlusProcessingContext alloc] init];
+  void *contentPointer = (void *)CFBridgingRetain(context);
+
+  @try {
+    [inputStream open];
+
+    hand = yajl_alloc(&callbacks, NULL, contentPointer);
+    yajl_config(hand, yajl_allow_comments, 0);
+    yajl_config(hand, yajl_dont_validate_strings, 1);
+
+    // Read json file
+    const NSUInteger inputBufferLength = 256;
+    uint8_t inputBuffer[inputBufferLength];
+    NSInteger read;
+
+    while ((read = [inputStream read:inputBuffer maxLength:inputBufferLength])) {
+
+      yajl_status status = yajl_parse(hand, inputBuffer, read);
+      if (status != yajl_status_ok) {
+        *error = [[self class] createParserError:hand];
+        return NO;
+      }
+    }
+
+    // Close parser
+    yajl_status status = yajl_complete_parse(hand);
+    if (status != yajl_status_ok) {
+      *error = [[self class] createParserError:hand];
+      return NO;
+    }
+  }
+  @catch (NSException *exception) {
+    *error = [NSError errorWithDomain:AdblockPlusErrorDomain code:0 userInfo:@{NSLocalizedDescriptionKey: [exception reason]}];
+    return NO;
+  }
+  @finally {
+    CFBridgingRelease(contentPointer);
+    [inputStream close];
+    yajl_free(hand);
+  }
+
+  // JSON is atleast valid
```

Comment:
I don't really understand that comment. JSON is "at least" valid? That sounds pessimistic. What I see happen here is, that we don't check for the presence of meta data fields and initialise them with defaults if it's a version 1 filter list (i.e. a plain content blocking list). Doesn't really need a comment how I see it, but if there is one it should probably explain that.